plan_monitor/collect.py

```diff
@@ -0,0 +1,154 @@
+
+
+def collect() -> None:
+    schema_registry = confluent_kafka.schema_registry.SchemaRegistryClient({'url': config.SCHEMA_REGISTRY_URL})
+    key_serializer = confluent_kafka.schema_registry.avro.AvroSerializer(
+        message_schemas.STATS_MESSAGE_KEY_AVRO_SCHEMA, schema_registry)
+    value_serializer = confluent_kafka.schema_registry.avro.AvroSerializer(
+        message_schemas.STATS_MESSAGE_VALUE_AVRO_SCHEMA, schema_registry)
+    producer_config = {'bootstrap.servers': config.KAFKA_BOOTSTRAP_SERVERS,
+                       'key.serializer': key_serializer,
+                       'value.serializer': value_serializer,
+                       'linger.ms': 100,
+                       'retry.backoff.ms': 250,
+                       'compression.codec': 'snappy'}
+    kafka_producer = confluent_kafka.SerializingProducer(producer_config)
+    result_queue = mp.Queue(10000)
+    stop_event = mp.Event()
+    produced_count = 0
+    processes = []
+
+    for db_identifier, odbc_conn_string in config.ODBC_CONN_STRINGS.items():
+        processes.append(mp.Process(target=poll_db, name=f'db-{db_identifier}',
+                                    args=(db_identifier, odbc_conn_string, stop_event, result_queue)))
+
+    start_time = time.perf_counter()
+
+    for process in processes:
+        process.start()
+        # Stagger so the herd doesn't dump a ton onto the results queue or Kafka producer at once:
+        time.sleep(0.5)
+
+    logger.info(f'Started {len(processes)} DB poll subprocesses')
+
+    try:
+        while not (stop_event.is_set() and result_queue.empty()):
+            kafka_producer.poll(0)  # serve delivery callbacks
+            try:
+                msg_value = result_queue.get(timeout=0.1)
+            except queue.Empty:
+                continue
+            msg_key = {"db_identifier": msg_value["db_identifier"],
+                       "set_options": msg_value["set_options"],
+                       "sql_handle": msg_value["sql_handle"]}
+            kafka_producer.produce(topic=config.STATS_TOPIC, key=msg_key, value=msg_value,
+                                   on_delivery=common.kafka_producer_delivery_cb)
+            produced_count += 1
+            if produced_count % 100_000 == 0:
+                logger.info(f"Produced {produced_count:,} stats records since process start...")
+                if not all([p.is_alive() for p in processes]):
+                    logger.warning('One or more subprocesses have failed in some sneaky way! Exiting...')
+                    stop_event.set()
+    except KeyboardInterrupt:
+        logger.info('Received interrupt request; shutting down...')
+    finally:
+        elapsed = (time.perf_counter() - start_time)
+        logger.info(f'Exiting after {produced_count:,} records were processed in {elapsed:.1f} seconds. Cleaning up...')
+        stop_event.set()
+        time.sleep(1)
+        result_queue.close()
+        result_queue.join_thread()
```

Comment:
what does joining here do?  I would have expected you to join on the individual processes first.  How are we consuming the rest of the things in the result_queue after we stop?

Reply:
The joining is now removed. As for what consumes the rest of the queue, that comes from the condition on the while loop at what is now L129.